src/SpecFetcher.php

```diff
@@ -0,0 +1,240 @@
+        ]);
+        return FALSE;
+      }
+
+      $data = (string) $response->getBody();
+      if (($file_size = $response->getBody()->getSize()) && $file_size < 1) {
+        $this->log(LogLevel::ERROR, 'API Doc %label: OpenAPI specification file located at %url is empty.', [
+          '%url' => $source_uri,
+          '%label' => $apidoc->label(),
+        ]);
+        return FALSE;
+      }
+
+      // Only save file if it hasn't been fetched previously.
+      $data_md5 = md5($data);
+      $prev_md5 = $apidoc->get('spec_md5')->isEmpty() ? NULL : $apidoc->get('spec_md5')->value;
+      if ($prev_md5 != $data_md5) {
+        $filename = $this->fileSystem->basename($source_uri);
+        $specs_definition = $apidoc->getFieldDefinition('spec')->getItemDefinition();
+        $target_dir = $specs_definition->getSetting('file_directory');
+        $uri_scheme = $specs_definition->getSetting('uri_scheme');
+        $destination = "$uri_scheme://$target_dir/";
+
+        try {
+          $this->checkRequirements($destination);
+          $file = file_save_data($data, $destination . $filename, FileSystemInterface::EXISTS_RENAME);
+
+          if (empty($file)) {
+            throw new \Exception('Could not save API Doc specification file.');
+          }
+        }
+        catch (\Exception $e) {
+          $this->log(LogLevel::ERROR, 'Error while saving API Doc spec file from URL on API Doc ID: %id. Error: %error', [
+            '%id' => $apidoc->id(),
+            '%error' => $e->getMessage(),
+          ]);
+          return FALSE;
+        }
+
+        $spec_value = ['target_id' => $file->id()] + $spec_value;
+        $apidoc->set('spec', $spec_value);
+        $apidoc->set('spec_md5', $data_md5);
+        $apidoc->set('fetched_timestamp', time());
+
+        return TRUE;
+      }
+    }
+
+    return FALSE;
+  }
+
+  /**
+   * Log a message, and optionally display it on the UI.
+   *
+   * @param string $level
+   *   The Error level.
+   * @param string $message
+   *   The message.
+   * @param array $params
+   *   Optional parameters array.
+   */
+  private function log(string $level, string $message, array $params = []) {
+    $this->logger->log($level, $message, $params);
+    // Show the message.
+    $this->messenger()->addMessage(new FormattableMarkup($message, $params), static::LOG_LEVEL_MAP[$level] ?? MessengerInterface::TYPE_ERROR);
+  }
+
+  /**
+   * Checks requirements for saving of a file spec.
+   *
+   * If a requirement is not fulfilled it throws an exception.
+   *
+   * @param string $destination
+   *   The specification file destination directory, including scheme.
+   *
+   * @throws \Exception
+   */
+  private function checkRequirements(string $destination): void {
+    // If using private filesystem, check that it's been configured.
+    if (strpos($destination, 'private://') === 0 && !$this->isPrivateFileSystemConfigured()) {
+      throw new \Exception('Private filesystem has not been configured.');
```

Comment:
https://github.com/apigee/apigee-edge-drupal/pull/191/files#r284212260
![image](https://user-images.githubusercontent.com/1755573/58001872-0e6e9500-7add-11e9-8b5a-0b8774e0624a.png)

Reply:
I have no problem with addressing this in a followup PR.

Reply:
Which is the related follow-up task to this? Probably it would be better to create it as a GH issue, because that could be referenced in this PR.